haskell/src/Stablecoin/Client/Cleveland/IO.hs

```diff
@@ -0,0 +1,166 @@
+import Lorentz.Contracts.Stablecoin (OwHook(..), OwHookOptReq(..), PermissionsDescriptor)
+import Stablecoin.Client (AddressAndAlias(..))
+
+----------------------------------------------------------------------------
+-- CLI
+----------------------------------------------------------------------------
+
+callStablecoinClient :: MorleyClientConfig -> [String] -> IO Text
+callStablecoinClient conf args = toText <$> do
+  let allArgs = args <> morleyClientOpts conf
+  readProcessWithExitCode "stablecoin-client" allArgs "N" >>=
+    \case
+      (ExitSuccess, output, errOutput) ->
+        output <$ putErrLn errOutput
+      (ExitFailure code, toText -> output, toText -> errOutput) ->
+        throwM $ MorleyClient.UnexpectedClientFailure code output errOutput
+
+-- | Write something to stderr.
+putErrLn :: Print a => a -> IO ()
+putErrLn = hPutStrLn stderr
+
+morleyClientOpts :: MorleyClientConfig -> [String]
+morleyClientOpts (MorleyClientConfig _ap addr port tzPath dataDir useHttps verb key) =
+  encodeMaybeOption "--node-url" addr
+  <> encodeMaybeOption "--node-port" port
+  <> ["--client-path", tzPath]
+  <> encodeMaybeOption "--data-dir" dataDir
+  <> if useHttps then ["--use-https"] else []
+  <> replicate (fromIntegral verb) "-V"
+  <> encodeMaybeOption "--secret-key" key
+
+encodeMaybeOption :: Buildable a => String -> Maybe a -> [String]
+encodeMaybeOption _ Nothing = []
+encodeMaybeOption optName (Just val) = [optName, pretty val]
+
+----------------------------------------------------------------------------
+-- Parse stdout
+----------------------------------------------------------------------------
+
+type Parser = P.Parsec OutputParseError Text
+
+data OutputParseError = OutputParseError Text Text
+  deriving stock (Eq, Show, Ord)
+
+instance Exception OutputParseError where
+  displayException = showErrorComponent
+
+instance ShowErrorComponent OutputParseError where
+  showErrorComponent (OutputParseError parserName err) =
+    "Failed parsing '" +| parserName |+ "\nError: " +| err |+ ""
+
+runParser :: Text -> Parser a -> IO a
+runParser str p =
+  case P.parse p "" str of
+    Right a -> pure a
+    Left err -> throwM err
+
+labelled :: Text -> Parser a -> Parser a
+labelled label p =
+  P.skipManyTill (printChar <|> newline) (string (label <> ": ")) >> p
+
+mutezParser :: Parser Mutez
+mutezParser =
+  -- Note: There is a known bug in `formatter` that makes `Mutez 0` be printed
+  -- as "-0 μꜩ" instead of "0 μꜩ", so we need to handle this special case here.
+  -- https://github.com/chrisdone/formatting/pull/61
+  -- https://gitlab.com/morley-framework/morley/-/issues/302#note_386133651
+  (string "-0 μꜩ" $> unsafeMkMutez 0)
+  <|>
+  (fmap unsafeMkMutez decimal <* string " μꜩ")
+
+naturalParser :: Parser Natural
+naturalParser = decimal
+
+textParser :: Parser Text
+textParser = fromString <$> some printChar
+
+mTextParser :: Parser MText
+mTextParser = either error id . mkMText <$> textParser
```

Comment:
I think it's better to fail using some `megaparsec`'s function rather than `error`. Though, since it's used only in tests and only to parse the output of `stablecoin-client`, I think `error` is acceptable as well, but then please add `HasCallStack` just in case.

Reply:
Done